RealmSwift/List.swift

```diff
@@ -508,14 +494,168 @@ extension List: RealmCollection, RangeReplaceableCollection {
+                block(RealmCollectionChange.fromObjc(value: anyCollection, change: change, error: error))
+            }
+    }
+}
+
+#if swift(>=4.0)
+// MARK: - MutableCollection conformance, range replaceable collection emulation
+extension List: MutableCollection {
+    public typealias SubSequence = RandomAccessSlice<List>
+
+    /**
+     Returns the objects at the given range (get), or replaces the objects at the
+     given range with new objects (set).
+
+     - warning: Objects may only be set during a write transaction.
+
+     - parameter index: The index of the object to retrieve or replace.
+     */
+    public subscript(bounds: Range<Int>) -> SubSequence {
+        get {
+            return SubSequence(base: self, bounds: bounds)
+        }
+        set {
+            replaceSubrange(bounds.lowerBound..<bounds.upperBound, with: newValue)
         }
     }
+
+    /**
+     Removes the first object in the list. The object is not removed from the
+     Realm that manages it.
+
+     This is a no-op if the list is already empty.
+
+     - warning: This method may only be called during a write transaction.
+     */
+    public func removeFirst() {
+        guard _rlmArray.count > 0 else {
+            return
+        }
+        _rlmArray.removeObject(at: 0)
+    }
+
+    /**
+     Removes the specified number of objects from the beginning of the list. The
+     objects are not removed from the Realm that manages them.
+
+     This is a no-op if the list is already empty.
+
+     - warning: This method may only be called during a write transaction.
+     */
+    public func removeFirst(_ number: Int) {
+        let count = Int(_rlmArray.count)
+        guard number <= count else {
+            throwRealmException("It is not possible to remove more objects from a list than it already contains.")
+            return
+        }
+        for _ in 0..<number {
+            _rlmArray.removeObject(at: 0)
+        }
+    }
+
+    /**
+     Removes the specified number of objects from the end of the list. The objects
+     are not removed from the Realm that manages them.
+
+     This is a no-op if the list is already empty.
+
+     - warning: This method may only be called during a write transaction.
+     */
+    public func removeLast(_ number: Int) {
+        let count = Int(_rlmArray.count)
+        guard number <= count else {
+            throwRealmException("It is not possible to remove more objects from a list than it already contains.")
+            return
+        }
+        for _ in 0..<number {
+            _rlmArray.removeLastObject()
+        }
+    }
+
+    /**
+     Inserts the items in the given collection into the list at the given position.
+
+     - warning: This method may only be called during a write transaction.
+     */
+    public func insert<C: Collection>(contentsOf newElements: C, at i: Int)
+        where C.Iterator.Element == Element {
+            for item in newElements.reversed() {
```

Comment:
`for item in newElements { insert(item, at: i); i += 1 }` is asymptotically faster, as inserting them in reverse order at `i` requires shifting all of the already-inserted values each time.